DirectProgramming/DPC++/N-BodyMethods/Nbody/src/GSimulation.cpp

```diff
@@ -0,0 +1,234 @@
+       h.parallel_for(r, [=](id<1> i) {
+         p[i].vel[0] += p[i].acc[0] * dt;  // 2flops
+         p[i].vel[1] += p[i].acc[1] * dt;  // 2flops
+         p[i].vel[2] += p[i].acc[2] * dt;  // 2flops
+
+         p[i].pos[0] += p[i].vel[0] * dt;  // 2flops
+         p[i].pos[1] += p[i].vel[1] * dt;  // 2flops
+         p[i].pos[2] += p[i].vel[2] * dt;  // 2flops
+
+         p[i].acc[0] = 0.f;
+         p[i].acc[1] = 0.f;
+         p[i].acc[2] = 0.f;
+
+         e[i] = p[i].mass *
+                (p[i].vel[0] * p[i].vel[0] + p[i].vel[1] * p[i].vel[1] +
+                 p[i].vel[2] * p[i].vel[2]);  // 7flops
+       });
+     })
+        .wait_and_throw();
+	// Third kernel accumulates the energy of this Nbody system
+    q.submit([&](handler& h) {
+       auto e = ebuf.get_access<access::mode::read_write>(h);
+       h.single_task([=]() {
+         for (int i = 1; i < n; i++) e[0] += e[i];
+       });
+     })
+        .wait_and_throw();
+    auto a = ebuf.get_access<access::mode::read_write>();
+    kenergy_ = 0.5 * a[0];
+    a[0] = 0;
+
+    double elapsed_seconds = ts0.Elapsed();
+    if (!(s % get_sfreq())) {
+      nf += 1;
+      std::cout << " " << std::left << std::setw(8) << s << std::left
+                << std::setprecision(5) << std::setw(8) << s * get_tstep()
+                << std::left << std::setprecision(5) << std::setw(12) << kenergy_
+                << std::left << std::setprecision(5) << std::setw(12)
+                << elapsed_seconds << std::left << std::setprecision(5)
+                << std::setw(12) << gflops * get_sfreq() / elapsed_seconds
+                << "\n";
+      if (nf > 2) {
+        av += gflops * get_sfreq() / elapsed_seconds;
+        dev += gflops * get_sfreq() * gflops * get_sfreq() /
+               (elapsed_seconds * elapsed_seconds);
+      }
+    }
+
+  }  // end of the time step loop
+  total_time_ = t0.Elapsed();
+  total_flops_ = gflops * get_nsteps();
+
+  av /= (double)(nf - 2);
+  dev = sqrt(dev / (double)(nf - 2) - av * av);
+
+  std::cout << "\n";
+  std::cout << "# Total Time (s)     : " << total_time_ << "\n";
+  std::cout << "# Average Performance : " << av << " +- " << dev << "\n";
+  std::cout << "===============================" << "\n";
+}
+
+/* Print the headers for the output */ 
+void GSimulation::PrintHeader() {
+  std::cout << " nPart = " << get_npart() << "; "
+            << "nSteps = " << get_nsteps() << "; "
+            << "dt = " << get_tstep() << "\n";
+
+  std::cout << "------------------------------------------------" << "\n";
+  std::cout << " " << std::left << std::setw(8) << "s" << std::left
+            << std::setw(8) << "dt" << std::left << std::setw(12) << "kenergy"
+            << std::left << std::setw(12) << "time (s)" << std::left
+            << std::setw(12) << "GFlops" << "\n";
+  std::cout << "------------------------------------------------" << "\n";
+}
+
+GSimulation::~GSimulation() { delete particles_; }
```

Comment:
This code was wrong from start because it should have been `delete[] particles_` anyway...

Reply:
No more destructors since the code uses std::vector instead of pointers now